pkg/tbtc/maintainer/bitcoin_difficulty.go

```diff
@@ -2,54 +2,271 @@ package maintainer
+	if err != nil {
+		return fmt.Errorf(
+			"cannot check whether genesis has been performed [%v]",
+			err,
+		)
+	}
+
+	if !isReady {
+		return fmt.Errorf(
+			"genesis has not been performed in the Bitcoin difficulty chain",
+		)
+	}
+
+	authorizationRequired, err := bdm.chain.AuthorizationRequired()
+	if err != nil {
+		return fmt.Errorf(
+			"cannot check whether authorization is required to submit "+
+				"block headers [%v]",
+			err,
+		)
+	}
+
+	if !authorizationRequired {
+		return nil
+	}
+
+	maintainerAddress := bdm.chain.Signing().Address()
+
+	isAuthorized, err := bdm.chain.IsAuthorized(maintainerAddress)
+	if err != nil {
+		return fmt.Errorf(
+			"cannot check whether Bitcoin difficulty maintainer is authorized "+
+				"to submit block headers [%v]",
+			err,
+		)
+	}
+
+	if !isAuthorized {
+		return fmt.Errorf(
+			"bitcoin difficulty maintainer has not been authorized to submit " +
+				"block headers",
+		)
+	}
+
+	return nil
+}
+
+// proveNextEpoch proves a single Bitcoin blockchain epoch in the Bitcoin
+// difficulty chain if there is a Bitcoin blockchain epoch to be proven.
+func (bdm *BitcoinDifficultyMaintainer) proveNextEpoch() error {
+	// The height of the Bitcoin blockchain.
+	currentBlockNumber, err := bdm.btcChain.GetCurrentBlockNumber()
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get current block number [%v]",
+			err,
+		)
+	}
+
+	// The current epoch proven in the Bitcoin difficulty chain.
+	currentEpoch, err := bdm.chain.CurrentEpoch()
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get current epoch [%v]",
+			err,
+		)
+	}
+
+	// The number of blocks required for each side of a retarget proof.
+	proofLength, err := bdm.chain.ProofLength()
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get proof length [%v]",
+			err,
+		)
+	}
+
+	// The new epoch to be proven in the Bitcoin difficulty chain.
+	newEpoch := currentEpoch + 1
+
+	// Height of the first block of the new epoch.
+	newEpochHeight := newEpoch * bitcoinEpochLength
+
+	// The range of block headers to be pull from the Bitcoin chain should
+	// start `proofLength` blocks before the first block of the new epoch
+	// and end `proofLength`-1 after it.
+	// For example, if the new epoch begins at block 522144 and `proofLength`
+	// is 3, then the range should be [522141, 522146]:
+	// 522141 <- old difficulty target
+	// 522142 <- old difficulty target
+	// 522143 <- old difficulty target
+	// << difficulty retarget >>
+	// 522144 <- new difficulty target (first block of the new epoch)
+	// 522145 <- new difficulty target
+	// 522146 <- new difficulty target
+	firstBlockHeaderHeight := uint(newEpochHeight - proofLength)
+	lastBlockHeaderHeight := uint(newEpochHeight + proofLength - 1)
+
+	// The required range of block headers can be pulled from the Bitcoin
+	// blockchain only if the blockchain height is equal to or greater than
+	// the end of the range.
+	if currentBlockNumber >= lastBlockHeaderHeight {
```

Comment:
One thing to note here is that after each Bitcoin epoch we will have to wait for a while before we can prove it.
For example, if `proofLength` is 20 blocks, we will have to wait 20 Bitcoin blocks (about 3 hours and 20 minutes) before the  new epoch is proven in the `LightRelay`.

If the wallet makes some transaction that requires a proof and the transaction is made shortly after the Bitcoin difficulty epoch is changed, we will not be able to prove it in the Bridge for 20 blocks. 

 This window lasting `proofLength` blocks will happen every ~2 weeks.
 Are we ok with that?

Reply:
Waiting for `proofLength` blocks after the new epoch's start seems to be the simplest idea.